Add support for conditions on the device tree machine compatible

ukleinek opened the pull request:
This is similar in spirit to ConditionArchitecture and checks the
content of /sys/firmware/devicetree/base/compatible against the passed
string. Compared to the Architecture this is more fine grained.

poettering commented:
hmm, i find this awfully specific I must say. Can't we find a more generic concept for this?

i.e. make this ConditionFirmware= that can accept various different flags, maybe ConditionFirmware=uefi, ConditionFirmware=bios, ConditionFirmware=device-tree and then ConditionFirmware=device-tree-compatible(xyz) or so as additional syntax for what you are looking for here?

ukleinek commented:
> hmm, i find this awfully specific I must say. Can't we find a more generic concept for this?

I thought about something more generic, too, but didn't find anything I was happy with.

> i.e. make this ConditionFirmware= that can accept various different flags, maybe ConditionFirmware=uefi, ConditionFirmware=bios, ConditionFirmware=device-tree and then ConditionFirmware=device-tree-compatible(xyz) or so as additional syntax for what you are looking for here?

Apart from now knowing how to test for the non-dt cases, I don't like the naming because "firmware" is a too generic concept (i.e. a wlan adapter has firmware, too).

Another not so nice thing is that e.g. `uefi` and `device-tree` are not orthogonal as UEFI is a system software still running in background(?) and device-tree is the description of the components of the machine. Something more in line with the device-tree type of condition is the machine id that used to identify the machine in ARM land. But as this is a deprecated thing I wouldn't add support for this to systemd today.

On my x86 laptop `dmidecode` has:
```
Handle 0x000C, DMI type 1, 27 bytes
System Information
        Manufacturer: LENOVO
        Product Name: 20FWCTO1WW
        Version: ThinkPad T460p
        Serial Number: PF0QG669
        UUID: 17c0a84c-234d-11b2-a85c-8a898c47e66c
        Wake-up Type: Power Switch
        SKU Number: LENOVO_MT_20FW_BU_Think_FM_ThinkPad T460p
        Family: ThinkPad T460p
```

Maybe `ConditionSystemFamily=` that behaves like my suggested `MachineDtCompatible=` on a dt-defined system and uses `Family` item in the dmi system information?

poettering commented:
Compare with ConditionSecurity=, the values accepted there aren't orthogonal either. i.e. there's "audit" and "selinux" and "uefi-secureboot", and typically you use all three. I see no conceptual problem at all with non-orthogonality at all here.

systemd is according to our docs "a system and service manager", and hence I would say if we just refer to "firmware" it should be clear enough we talk about system firmware, no?

I like that "system firmware" is pretty generic, we could then also use it for dmi string matches and whatnot if we wanted. i.e. consider "ConditionFirmware=dmi-vendor(Lenovo)" or so?

poettering commented:
> Apart from now knowing how to test for the non-dt cases,

Well, don#t add what you can't test, it's totally OK if it initially just contains checks for generic stuff, i.e. maybe "ConditionFirmware=device-tree" to check whether devicetree is available ad all, and "ConditionFirmware=device-tree-compatible(xyz)" for your compatible thing? it's totally ok if that's all for now.

(checking for uefi is trivial btw, just use is_efi_boot() from src/basic/efivars.h)

poettering commented:
(no idea how to check for a classic PC bios though)

ukleinek commented:
for the record: I use this patch locally and there I noticed that there is a memory management bug in my patch. So if someone wants to work on this (maybe a future me): make sure not to modify `buf` in the function as this is autofree'd

ukleinek commented:
I reworked the patch to work on the syntax that @poettering suggested now and rebased to current main.

One not-so-nice thing is that the error messages don't refer to the unit that triggered the error. But that is also the case for the other conditions, so this is orthogonal to this patch.

ukleinek commented:
force-pushed to fix a compiler warning

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+        assert(c->parameter);
+        assert(c->type == CONDITION_FIRMWARE);
+
+        log_error("%s:%d\n", __func__, __LINE__);
```
i think you forgot to erase some debug line here

ukleinek replied:
fixed now

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+
+        log_error("%s:%d\n", __func__, __LINE__);
+        if (!strcmp(c->parameter, "uefi")) {
+                return is_efi_boot();
```
as per https://systemd.io/CODING_STYLE/#formatting no {} around single-line if blocks, please

ukleinek replied:
Ah, even if an else branch is multi-line?

poettering replied:
yupp, even then

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+        assert(c->type == CONDITION_FIRMWARE);
+
+        log_error("%s:%d\n", __func__, __LINE__);
+        if (!strcmp(c->parameter, "uefi")) {
```
we prefer use of streq() for string equality checks. It's more readable. It's just a macro that is equivalent to strcmp() == 0, see coding style

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+        log_error("%s:%d\n", __func__, __LINE__);
+        if (!strcmp(c->parameter, "uefi")) {
+                return is_efi_boot();
+        } else if (!strcmp(c->parameter, "device-tree")) {
```
as above: streq(), and no {}

ukleinek replied:
as above, it's even the same code line :-)

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+        if (!strcmp(c->parameter, "uefi")) {
+                return is_efi_boot();
+        } else if (!strcmp(c->parameter, "device-tree")) {
+                return access("/sys/firmware/device-tree/", F_OK) >= 0;
```
for new code we prefer if unexpected errors are logged at debug level even if ignore, i.e. maybe rewrite to:

```c
if (access("/sys/firmware/device-tree/", F_OK) < 0) {
        if (errno != ENOENT)
                log_debug_errno("Unexpected error when checking for /sys/firmware/device-tree/: %m");
        return false;
} else
        return true;
```

or so

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+                return is_efi_boot();
+        } else if (!strcmp(c->parameter, "device-tree")) {
+                return access("/sys/firmware/device-tree/", F_OK) >= 0;
+        } else if (!strncmp(c->parameter, "device-tree-compatible(", strlen("device-tree-compatible("))) {
```
startswith()

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+                _cleanup_free_ char *dtcompatarg = NULL;
+                char *end;
+
+                dtcompatarg = strdup(c->parameter + strlen("device-tree-compatible("));
```
use the return value of the startswith() call suggested above, which will point to the first char after the prefix. Then use endswith() to get a pointer to the the closing bracket, and cut out what's between with strndup()

ukleinek replied:
I don't like endswith() here, because I want to refuse
    Firmware=device-tree-compatible(some)string)
But I got your point.

poettering replied:
good point

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+                dtcompatarg = strdup(c->parameter + strlen("device-tree-compatible("));
+                if (!dtcompatarg) {
+                        log_debug("Failed to allocate memory, assuming machine is incompatible");
+                        return false;
```
please propagate OOM as `return -ENOMEM;` up

i.e. it's OK to sometimes eat up errors (and debug log about that), but OOM is usually not among those cases, and it should always be propagated up

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+        }
+
+        /* Not sure this can happen, but play safe. */
+        if (!size) {
```
we usually use C's downgrade to bool feature only for variables that actually are bools or at least boolean like – and for pointers, but never for values that are actually numbers, like this one. Hence please write as `size == 0`

poettering replied:
not addressed

ukleinek replied:
Oh, indeed, I failed to fix this. Fixed in the most recent push.

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
+
+                size -= compat_len;
+                compat += compat_len;
+        } while (size);
```
maybe just do:

```c
l = strv_parse_nulstr(dtcompat, size);
if (!l)
        return -ENOMEM;
return strv_contains(l, dtcomparg);
```

i.e. use the parser we already have for this. yes, this will do more memory allocations, but that shouldn't really matter, this is not performance critical after all, and I much prefer if we'd reuse parsers as much as possible, since they are after all problematic for security.

ukleinek commented:
The failing checks are unrelated to my changes, they all seem to fail because salsa.debian.org is/was unavailable.

ukleinek commented:
I just pushed just the same tree again (with a newer committer date) to make the tests rerun. Hopefully the Debian tests don't fail this time. Maybe this was the only problem before this pull request is considered again for review?!

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,81 @@ static int condition_test_architecture(Condition *c, char **env) {
 
+#define DTCOMPAT_FILE "/sys/firmware/devicetree/base/compatible"
+static int condition_test_firmware_devicetree_compatible(const char *dtcarg) {
+        int err;
```
doesn't matter, but throughout the codebase we use "r" as the name for the variable we store function call return values into. Please do the same here, i.e. just rename `err` to `r`

poettering commented on src/shared/condition.c:
```diff
@@ -450,6 +451,81 @@ static int condition_test_architecture(Condition *c, char **env) {
+        if (err < 0) {
+                /* if the path doesn't exist it is incompatible */
+                if (err != -ENOENT)
+                        log_debug_errno(errno, "Failed to open() '%s', assuming machine is incompatible: %m", DTCOMPAT_FILE);
```
the error code is in `err`, hnece please pass that as first arg to log_debug_errno()

ukleinek commented:
Among the four failed tests are two that fail for reasons unrelated to my change ("Build test/ build (clang, 11) (pull request)" and "Build test / build (clang, 12) (pull_request)" both failed to install the toolchain with apt errors.)

I don't understand the two others and doubt that they are related to my changes. Are these know to be unstable?

ukleinek commented:
(Just pushed an amended commit (i.e. the same changes, just a newer committer-date) to trigger the CI checks. I guess they were false-negative last time again.)

ukleinek commented:
OK, the test failed for a good reason and now I managed to interpret the failure. I hope I fixed it with the new force-push now.

ukleinek commented:
I wonder why this pull request doesn't get further attention. Is it -ENOTIME (or -ENOINTERST?) for maintainers? Or did if fell through the cracks?
